library/fsio.c

```diff
@@ -0,0 +1,395 @@
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+
+#include <string.h>
+#include <assert.h>
+#include "mbedtls/fsio.h"
+
+#if !defined(MBEDTLS_FS_IO_ALT)
+
+#if !defined(MBEDTLS_PLATFORM_NO_STD_FUNCTIONS)
+
+#include <sys/stat.h>
+
+/**
+ * \brief          Open file and disable buffering.
+ *
+ * \param path     File path
+ * \param mode     Open mode
+ *
+ * \return         File handle of type mbedtls_file_t.
+ *                 On failure MBEDTLS_FILE_INVALID is returned.
+ */
+mbedtls_file_t mbedtls_fopen( const char *path, const char *mode )
+{
+    mbedtls_file_t f = fopen( path, mode );
+
+    if( f != MBEDTLS_FILE_INVALID && ( setvbuf( f, NULL, _IONBF, 0 ) != 0 ) )
+    {
+        fclose( f );
+        f = MBEDTLS_FILE_INVALID;
+    }
+
+    return( f );
+}
+
+/**
+ * \brief           Read next directory entry.
+ *
+ * \note            This is rework of readdir to adapt to abstraction. This does
+ *                  not return struct dirent pointer. Hence no need to worry
+ *                  pointer ownership and cleanup.
+ *
+ * \param dir       Dir handle of type mbedtls_dir_t.
+ * \param file_name Out buffer for directory entry name.
+ *                  Up to 255 character long name can be returned.
+ *                  Returned string is NULL terminated.
+ * \param size      Out buffer length.
+ *
+ * \return          0 for success. -1 for failure.
+ */
+int mbedtls_readdir( mbedtls_dir_t dir, char *file_name, uint32_t size )
+{
+    int status = -1;
+    struct dirent *entry;
+
+    if ( ( entry = readdir( dir ) ) != NULL )
+    {
+        strncpy( file_name, entry->d_name, size );
+        if ( file_name[size - 1] == '\0' ) /* Check if buffer was enough */
+        {
+            status = 0;
+        }
+        else
+        {
+            file_name[0] = 0;
+        }
+    }
+
+    return( status );
+}
+
+/**
+ * \brief          Get file stats.
+ *
+ * \param path     File path string
+ * \param sb       Output mbedtls_stat_t struct.
+ *
+ * \return         Returns 0 on success, -1 on failure.
+ */
+int mbedtls_stat( const char * path, mbedtls_stat_t * msb )
+{
+    int status = -1;
+    struct stat sb;
+
+    status = stat( path, &sb );
+    if ( status == 0 )
+    {
+        switch ( sb.st_mode & S_IFMT )
+        {
+            case S_IFREG:
```

Comment:
Also `S_ISFIFO` (pipes, very useful), `S_ISSOCK` (sockets, occasionally useful), `S_ISCHR` (for `/dev/fd/NNN` on BSD) and `S_ISBLK` (probably not useful in practice but technically correct).

Reply:
In context of Mbed TLS would these be useful?

Reply:
@mazimkhan Yes, as I wrote, pipes and character devices are useful because they come up in scripts. Pipes come up whenever you write something like `generate_key | consume_key`. Character devices come up with more complex redirections on *BSD because that' what `/dev/fd/*` are, and also when running under `expect`.

Reply:
Added in commit https://github.com/ARMmbed/mbedtls/pull/1029/commits/3e4dd8c5ac3981ff3a89d1633b08ebf2957327a6